runtime/compiler/env/j9method.cpp

```diff
@@ -4624,6 +4634,25 @@ TR_ResolvedJ9Method::TR_ResolvedJ9Method(TR_OpaqueMethodBlock * aMethod, TR_Fron
             setRecognizedMethodInfo(TR::java_util_ArrayList_all);
          else if ((classNameLen == 19) && !strncmp(className, "java/util/Hashtable", 19))
             setRecognizedMethodInfo(TR::java_util_Hashtable_all);
+         else if ((classNameLen == 21) && !strncmp(className, "java/lang/StringUTF16", 21))
+            {
+            // StringUTF16 has two methods that start with 'indexOf'. For vector support, we only
+            // recognize and inline indexOf().
+            if ((strncmp(name, "indexOf", 7) == 0) && (strncmp(name, "indexOfLatin1", 13) != 0))
+               setRecognizedMethodInfo(TR::java_lang_StringUTF16_indexOf);
+            }
```

Comment:
Can you explain why exactly we need this chunk of code and the subsequent `else if` blocks that were added?

Reply:
This new checks are all under `      if (TR_Method::getMandatoryRecognizedMethod() == TR::unknownMethod)` where multiple methods all have the same signature and very similar names. The OpenJ9 recognized method mechanism could not distinguish Java functions like these.

Take the `StringUTF16` class for example, it has these two methods:
```
    public static int indexOf(byte[] value, int valueCount, byte[] str, int strCount, int fromIndex) 
    public static int indexOfLatin1(byte[] src, int srcCount, byte[] tgt, int tgtCount, int fromIndex) 
```

They have the same signature and the names all start with `indexOf`.  If this `if check` were not added, the JIT would recognize both `indexOf` and `indexOfLatin1` as the same method. I've seen this incorrect behavior in a method trace and added this check to fix that.

The JITHelper recognized method checks are added for the same reason.

Reply:
Ok I think I understand. Why do we need to check for `"java/lang/StringLatin1"` below then? It doesn't have these duplicates. Also why do we need to check `"com/ibm/jit/JITHelpers"` it also doesn't have duplicates.